tracing-subscriber/src/filter/env/field.rs

```diff
@@ -36,14 +36,23 @@ pub(crate) struct MatchVisitor<'a> {
     inner: &'a SpanMatch,
 }
 
-#[derive(Debug, Clone, PartialOrd, Ord, Eq, PartialEq)]
+#[derive(Debug, Clone, PartialOrd, PartialEq)]
 pub(crate) enum ValueMatch {
     Bool(bool),
+    F64(f64),
     U64(u64),
     I64(i64),
     Pat(Box<MatchPattern>),
 }
 
+impl Eq for ValueMatch {}
+
+impl Ord for ValueMatch {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).unwrap()
+    }
+}
```

Comment:
This isn't really correct, because there's nothing stopping the `ValueMatch::F64` case from containing `NaN` values, and [`NaN`s don't have a total order](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html#how-can-i-implement-partialord). 

The total ordering is required because `ValueMatch`es are stored in a `BTreeMap`, so we can't just remove the `Ord` implementation. Instead, what I think we should do is make some additional changes to ensure that `ValueMatch::F64` variants will never contain `NaN` values, as I discussed in other comments. I think if we make those changes, we can then implement `Ord` for `ValueMatch` if we add an assertion when comparing `ValueMatch::F64` variants that check that they aren't `NaN` (or `expect` the result of the `partial_cmp` comparison). For every other variant, the `Ord` impl can just call that type's `Ord::cmp`.

I would expect the implementation to look something like this:
```suggestion
impl Ord for ValueMatch {
    fn cmp(&self, other: &Self) -> Ordering {
        use ValueMatch::*;
        match (self, other) {
            (Bool(this), Bool(that)) => this.cmp(that),
            (Bool(_), _) => Ordering::Less,
            
            (F64(this), F64(that)) => this.partial_cmp(that).expect("`ValueMatch::F64` may not contain `NaN` values"),
            (F64(_), Bool(_)) => Ordering::Greater,
            (F64(_), _) => Ordering::Less,
            
            (NaN, NaN) => Ordering::Equal,
            (NaN, Bool(_)) | (NaN, F64(_)) => Ordering::Greater,
            (NaN _) => Ordering::Less,
            
            (U64(this), U64(that)) => this.cmp(that),
            (U64(_), Bool(_)) | (U64(_), F64(_)) | (U64(_), NaN) => Ordering::Greater,
            (U64(_), _) => Ordering::Less,
            
            (I64(this), I64(that)) => this.cmp(that),
            (I64(_), Bool(_)) | (I64(_), F64(_)) | (I64(_), NaN), (I64(_), U64(_)) => Ordering::Greater,
            (I64(_), _) => Ordering::Less,
            
            (Pat(this), Pat(that)) => this.cmp(that),
            (Pat(_), _) => Ordering::Greater,
        }
    }
}

impl PartialOrd for ValueMatch {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}
```
or similar.

Reply:
:+1: